### ISSUE TITLE
Can't load samples from instrument libraries

### ISSUE THREAD

**Issue by KynikossDragonn** [schismtracker/schismtracker#370]
Something I can normally do just fine in Impulse Tracker and OpenMPT doesn't seem to want to work in Schism Tracker unless I'm completely missing something? Has this ever worked in Schism Tracker at any point?

I've built from source from this commit: https://github.com/schismtracker/schismtracker/commit/d5153477385756c98299d3f3260482ccf226d613

All I get is this:
![Screenshot_2023-08-03_10-52-54](https://github.com/schismtracker/schismtracker/assets/8813106/b91ac6ed-ba31-4237-9e6a-76db3c5bd1b7)

and if I try to select anything:
![Screenshot_2023-08-03_10-53-08](https://github.com/schismtracker/schismtracker/assets/8813106/8c0c87bd-1e84-47e9-940d-468aefab6133)

It shows up as "Gravis Patch File" as the format, but it just says "No sample"?

**Comment by vcaputo** [comment]
I've never tried it myself, but the format seems to be implemented.

Can you attach a .pat file you've tried and failed to use?

**Comment by KynikossDragonn** [comment]
I'm basically trying to use the PAT files that were both from the original Gravis Ultrasound set, and the "Pro Patches Lite" set, and none of them are able to be loaded as a sample.

It's worth noting sometimes PAT's can contain multiple samples and Impulse Tracker will recognize this and let you pick which sample you actually want, treating that particular PAT file as a "sample library" in that case.
[gravispat.zip](https://github.com/schismtracker/schismtracker/files/12252836/gravispat.zip)

**Comment by vcaputo** [comment]
Can at least confirm it does the same for me here.

Maybe one of the other maintainers has more experience with this part of schism and knows if it ever worked historically.

There's definitely .pat loader code, and some commits speak to fixing it at times, so something must make use of it.

**Comment by mrpapersonic** [comment]
For the record, loading individual samples from instruments has always been broken in schism, you can't load them from XI or even ITI files.
https://github.com/schismtracker/schismtracker/blob/d5153477385756c98299d3f3260482ccf226d613/schism/audio_loadsave.c#L1399
I'm not too sure how to go about implementing this. Looking at the directory "emulation" code I can see why someone would hate attempting to make this work...

**Comment by KynikossDragonn** [comment]
> For the record, loading individual samples from instruments has always been broken in schism, you can't load them from XI or even ITI files.
> 
> https://github.com/schismtracker/schismtracker/blob/d5153477385756c98299d3f3260482ccf226d613/schism/audio_loadsave.c#L1399
> 
> I'm not too sure how to go about implementing this. Looking at the directory "emulation" code I can see why someone would hate attempting to make this work...

I guess it doesn't help that Impulse Tracker itself was all ASM source code and a complete nightmare to try to decipher.

But it sounds like, I'd have to switch to Instrument control, load my PAT under Instruments, delete the instrument and keep the resulting sample and switch back to Sample control? I haven't tried that yet but that would be quite the tedious hackish workaround for this problem in the interim.

**Comment by mrpapersonic** [comment]
https://github.com/mrpapersonic/schismtracker/tree/smp-load-from-inst I've made a commit in this branch with experimental support for this, if you want to try it out.

**Comment by KungFuFurby** [comment]
Linking this to #159, as I recognize this from many, many years ago (however, #159 had a not a directory error: that error is gone, but the problem is not yet resolved). Built from source from commit ab2161da8050b101a649dfbdc99ed8d0683ae745 on a 64-bit Intel Mac on Mac OS X 10.15.7 to test it out, and...

...I got a "(null): Cannot allocate memory" error on the message log viewer when trying to extract individual samples. Tested against .PAT, .XI and .ITI files. They loaded fine as instruments. See #381.
